File: Searching algos/Binary-Search.py
#This is a Python 3 program for Binary Search using recursive method.

#In this program, it will return the index of the number(Which we are searching for).
#If number not present it will return -1.

#Given Array of Numbers

array = [1, 3, 5, 13, 14, 15, 23, 19, 30, 35]

#Element to be Searched
search_element = int(input("Enter the Number To Be Searched: "))


def binary_search(array, low, high, search_element):
    #condition for checking base case

    if (high >=low):
        mid = (high + low) // 2

        #if the element is present at the middle itself
        if (array[mid] == search_element):
            return mid

        #if the element is smaller than mid, then the number only be present in left subarray.
        elif array[mid] > search_element:
            return binary_search(array, low, mid - 1, search_element)

        #else the element can be present only in right subarray
        else:
            return binary_search(array, mid + 1, high, search_element)

    else:
        # Else element not present in the array.
        return -1

result = binary_search(array, 0, len(array)-1, search_element)

if (result != -1):
    print(f"Element is present at index {result}")
else:
    print("Element is not present in the given array")

File: Searching algos/BinarySearch.java
public class BinarySearch {
    
    
    public static void main(String[] args) {
        //IMPORTANT - array MUST be sorted for search to work
        int[] arrayToSearch = new int[]{1,4,5,6,7,10,14,15,21,34,33};
        System.out.println(iterativeBinarySearch(arrayToSearch, 6));
        System.out.println(recursiveBinarySearch(arrayToSearch, 15, 0, arrayToSearch.length - 1));
    }

    //iterative binary search
    public static int iterativeBinarySearch(int[] array, int valueToSearch) {
        //set your indexes for array traversal (beginning and end)
        int low = 0;
        int high = array.length - 1;

        int index = Integer.MAX_VALUE;
        
        //loop through array
        while (low <= high) {

            //set middle index to start search
            int middle = low  + ((high - low) / 2);

            //if value at index is less than value we are searching for, increment index down
            if (array[middle] < valueToSearch) {
                low = middle + 1;

            //if value at index is greater than value we are searching for, increment index up
            } else if (array[middle] > valueToSearch) {
                high = middle - 1;

            //here we have found the index of the value and can return 
            } else if (array[middle] == valueToSearch) {
                index = middle;
                break;
            }
        }
        return index;
    }


    //for recursive search
    public static int recursiveBinarySearch(int[] array, int valueToSearch, int low, int high) {
        //set middle index to start search
        int middle = low  + ((high - low) / 2);
            
        //if highest index becomes less than lowest index, value does not exist
        //breaks infinite recursion
        if (high < low) {
            return -1;
        }

        //return index if value matches
        if (valueToSearch == array[middle]) {
            return middle;
        
        //if value is less than middle index, call function again and search through left side of array
        } else if (valueToSearch < array[middle]) {
            return recursiveBinarySearch(array, valueToSearch, low, middle - 1);

        //if value is greater than middle index, call function again and search through right side of array
        } else {
            return recursiveBinarySearch(array, valueToSearch, middle + 1, high);
        }
    }
}

